### PR TITLE
36 get notifications

### PR DESCRIPTION
**Pull request by ihor-mk**

### CODE REVIEW

File: frontend/src/app/shared/components/user-notification/user-notification.component.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        if (storedUserData) {
+            userEmail = JSON.parse(storedUserData).email.toLowerCase();
+        }
+
+        this.notificationsHub.listenMessages((msg) => {
+            const broadcastMessage: INotification = JSON.parse(msg.toLowerCase());
+
+            if (broadcastMessage.email === userEmail.toLowerCase()) {
+                this.notifications.push(broadcastMessage);
+            }
```
Comment (Anatolii-Ivanov) [inline frontend/src/app/shared/components/user-notification/user-notification.component.ts:45]: It's better to implement in the way, that you will pass parameters on connection start and then send from backend only to those users, who needs this

**Review by Anatolii-Ivanov — CHANGES_REQUESTED** [review]

File: backend/EasySpeak.Notifier/EasySpeak.Notifier.WebAPI/Hubs/NotificationHub.cs [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+using EasySpeak.Core.Common.DTO;
+using EasySpeak.RabbitMQ.Interfaces;
+
+namespace EasySpeak.Notifier.WebAPI.Hubs
+{
+
+    public class NotificationHub : Hub
+    {
+        
+    }
```
[the remark this answers is not in the document]
Comment (ihor-mk) [reply backend/EasySpeak.Notifier/EasySpeak.Notifier.WebAPI/Hubs/NotificationHub.cs:11]: I use it to create a hub only, and then get context

**Review by Anatolii-Ivanov — DISMISSED** [review]

File: backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/NotificationService.cs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+                        {
+                            Id = notify.Id,
+                            Text = notify.Text,
+                            Type = notify.Type
+                        })
+                .ToListAsync();
+
+        public async Task<NotificationDto> CreateNotificationAsync(NotificationDto notificationDto)
+        {
+            var user = await _context.Users.Where(u => u.Id == _firebaseAuthService.UserId).FirstOrDefaultAsync();
```
Comment (tatianahutii) [inline backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/NotificationService.cs:36]:
```suggestion
            var user = await _context.Users.FirstOrDefaultAsync(u => u.Id == _firebaseAuthService.UserId);
```

File: backend/EasySpeak.Notifier/EasySpeak.Notifier.WebAPI/EasySpeak.Notifier.WebAPI.csproj [hunk 4]
```diff
@@ -12,6 +12,7 @@
   </ItemGroup>
 
   <ItemGroup>
+    <ProjectReference Include="..\..\EasySpeak.Core\EasySpeak.Core.Common\EasySpeak.Core.Common.csproj" />
```
Comment (tatianahutii) [inline backend/EasySpeak.Notifier/EasySpeak.Notifier.WebAPI/EasySpeak.Notifier.WebAPI.csproj:15]: Please add this to Dockerfile

File: backend/EasySpeak.Notifier/EasySpeak.Notifier.WebAPI/Program.cs [hunk 5]
```diff
@@ -23,6 +24,15 @@
 builder.Services.AddHostedService<ConsumerHostedService>();
 builder.WebHost.UseUrls("http://*:5070");
 
+builder.Services.AddCors(options =>
+{
+    options.AddPolicy("CorsPolicy", builder => builder
+        .WithOrigins("http://localhost:4200")
```
Comment (tatianahutii) [inline backend/EasySpeak.Notifier/EasySpeak.Notifier.WebAPI/Program.cs:30]:
```suggestion
        .WithOrigins("http://*:4200")
```

File: backend/EasySpeak.Notifier/EasySpeak.Notifier.WebAPI/Services/ConsumerHostedService.cs [hunk 6]
```diff
@@ -27,10 +33,12 @@ public override async Task StopAsync(CancellationToken cancellationToken)
 
         protected override Task ExecuteAsync(CancellationToken stoppingToken)
         {
+
             try
                 {
-                    _consumer.Recieve<string>((data) =>
+                    _consumer.Recieve<NewNotificationDto>((data) =>
                     {
+                        _hubContext.Clients.All.SendAsync($"Notification_{data.Email.ToLower()}", JsonConvert.SerializeObject(data));
```
Comment (tatianahutii) [inline backend/EasySpeak.Notifier/EasySpeak.Notifier.WebAPI/Services/ConsumerHostedService.cs:41]: please fix github actions warnings

File: frontend/src/app/core/hubs/notifications-hub.service.ts [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+export class NotificationsHubService {
+    readonly hubUrl = 'notificationsHub';
+
+    private hubConnection: HubConnection;
+
+    public readonly messages = new Subject<string>();
+
+    private subscriptions: Subscription[] = [];
+
+    // eslint-disable-next-line no-empty-function
```
Comment (tatianahutii) [inline frontend/src/app/core/hubs/notifications-hub.service.ts:19]: please do not use eslint-disable-next-line

File: frontend/src/app/shared/components/user-notification/user-notification.component.html [hunk 8]
```diff
@@ -0,0 +1,12 @@
+<div *ngIf="notifications.length > 0">
```
Comment (tatianahutii) [inline frontend/src/app/shared/components/user-notification/user-notification.component.html:1]:
```suggestion
<div *ngIf="notifications.length">
```

File: frontend/src/app/shared/components/user-notification/user-notification.component.ts [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+        this.notifySubscription = await this.httpService.get<INotification>('/notification').subscribe((data) => {
+            this.notifications = data;
+        });
+
+        await this.notificationsHub.start();
+
+        this.notificationsHub.listenMessages((msg) => {
+            const broadcastMessage: INotification = JSON.parse(msg.toLowerCase());
+
+            this.notifications.push(broadcastMessage);
```
Comment (tatianahutii) [inline frontend/src/app/shared/components/user-notification/user-notification.component.ts:36]: please get rid of mutations

**Review by tatianahutii — CHANGES_REQUESTED** [review]

File: backend/EasySpeak.Notifier/EasySpeak.Notifier.WebAPI/EasySpeak.Notifier.WebAPI.csproj (hunk 4 above)
Comment (tatianahutii) [reply backend/EasySpeak.Notifier/EasySpeak.Notifier.WebAPI/EasySpeak.Notifier.WebAPI.csproj:15]: still need this

Comment (ihor-mk) [reply backend/EasySpeak.Notifier/EasySpeak.Notifier.WebAPI/EasySpeak.Notifier.WebAPI.csproj:15]: This reference is present in Dockerfile

**Review by Anatolii-Ivanov — DISMISSED** [review]

**Review by NLMOTS — APPROVED** [review]
If using http necessarily, keep this like that

**Review by Anatolii-Ivanov — DISMISSED** [review]

**Review by NLMOTS — DISMISSED** [review]

**Review by Anatolii-Ivanov — APPROVED** [review]

**Review by NLMOTS — APPROVED** [review]
http error solved, good work

**Pull request merged**